routing/index_router.cpp

```diff
@@ -154,37 +154,60 @@ void PushPassedSubroutes(Checkpoints const & checkpoints, vector<Route::Subroute
 namespace routing
 {
 // IndexRouter::BestEdgeComparator ----------------------------------------------------------------
-IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point, m2::PointD const & direction)
-  : m_point(point), m_direction(direction)
+IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point,
+                                                    m2::PointD const & direction,
+                                                    IsOnewayFn const & isOnewayFn)
+  : m_point(point), m_direction(direction), m_isOnewayFn(isOnewayFn)
 {
 }
 
 bool IndexRouter::BestEdgeComparator::Compare(Edge const & edge1, Edge const & edge2) const
 {
-  if (!m_direction.IsAlmostZero())
+  if (IsDirectionValid())
   {
-    bool const isEdge1Parallel = IsAlmostParallel(edge1);
-    if (isEdge1Parallel != IsAlmostParallel(edge2))
+    bool const isEdge1Parallel = IsAlmostConformed(edge1);
+    if (isEdge1Parallel != IsAlmostConformed(edge2))
       return isEdge1Parallel;
   }
 
   return GetSquaredDist(edge1) < GetSquaredDist(edge2);
 }
 
-bool IndexRouter::BestEdgeComparator::IsAlmostParallel(Edge const & edge) const
+bool IndexRouter::BestEdgeComparator::IsAlmostConformed(Edge const & edge) const
 {
-  CHECK(!m_direction.IsAlmostZero(), ());
+  bool forward;
+  if (!IsAlmostCollinear(edge, forward))
+    return false;
+
+  // Note. In case of oneway edges |edge| is conformed to |m_direction| iff vector of |edge|
+  // and vector |m_direction| are almost collinear and have the same direction.
+  return m_isOnewayFn(edge.GetFeatureId()) ? forward : true;
+}
 
+bool IndexRouter::BestEdgeComparator::IsAlmostCollinear(Edge const & edge, bool & forward) const
+{
+  CHECK(IsDirectionValid(), ());
+
+  forward = true;
   m2::PointD const edgeDirection =
       edge.GetEndJunction().GetPoint() - edge.GetStartJunction().GetPoint();
   if (edgeDirection.IsAlmostZero())
     return false;
 
   double const cosAng =
     m2::DotProduct(m_direction.Normalize(), edgeDirection.Normalize());
+
   // Note. acos(0.97) ≈ 14 degrees.
   double constexpr kMinCosAngForAlmostParallelVectors = 0.97;
-  return  cosAng >= kMinCosAngForAlmostParallelVectors;
+  if (cosAng >= kMinCosAngForAlmostParallelVectors)
+    return true;
+
+  if (cosAng <= -kMinCosAngForAlmostParallelVectors)
```

Comment:
В этом случае оба будут коллинеарны направлению |m_direction|. Т.е. вызов IsAlmostCollinear(Edge const & edge, bool & forward) вернет true для обоих и заполнит forward разными значениями для каждого. Что логично, на мой взгляд. 

Далее будет использован метод:
```
bool IndexRouter::BestEdgeComparator::IsAlmostConformed(Edge const & edge) const
{
  bool forward;
  if (!IsAlmostCollinear(edge, forward))
    return false;

  // Note. In case of oneway edges |edge| is conformed to |m_direction| iff vector of |edge|
  // and vector |m_direction| are almost collinear and have the same direction.
  return m_isOnewayFn(edge.GetFeatureId()) ? forward : true;
}
```
Оба кандидата коллинеарны, поэтому рассматриваем только нижний retrun в коде выше. Для односторонних фичь он вернет true если вектор движения (m_direction) по ходу фичи (это верно, поскольку у нас нет реверсивных односторонних фич). А для кандидатов, лежащих на двусторонних фичах он вернет true в любом случае. Поскольку по двусторонней фичи можно двигаться в любую сторону.

Для bestEdge направление будет получено иначе (!). Не из кандидата, который обрабатывается в IsAlmostConformed(). Видимо в этом и проблема. Т.е. получается, что IsAlmostConformed() говорит, что (двусторонняя) дуга подходит по направлению (не зависимо от значения Edge::m_forward). А потом берется все из этой (двусторонней) дуги(Edge), а m_forward заменяется на правильный (в IndexRouter::FindBestSegment() при помощи getSegmentByEdge).

Думаю, лучше будет переписать ```bool IndexRouter::BestEdgeComparator::IsAlmostConformed(Edge const & edge) const``` с тем расчетом, чтоб для двусторонних дуг возвращалось true, только если они подходят (с учетом m_forward), и потом не подменять m_forward у хорошего двустороннего кандидата в IndexRouter::FindBestSegment().

Данное улучшение не повлияет на поведение, но упростит код и сделает его понятнее.